File: binary_tree/binary_tree_preorder_traversal.144.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// Iterative DFS using an explicit stack
// O(n) time
// O(n) space
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        if (!root) return {};
        std::vector<int> preorder;
        std::stack<TreeNode*> st({root});
        // std::stack<TreeNode*> st;
        // st.push(root);
        while (!st.empty()) {
            TreeNode* node = st.top();
            st.pop();
            preorder.push_back(node->val);
            if (node->right != nullptr) st.push(node->right);
            if (node->left != nullptr) st.push(node->left);
        }
        return preorder;
    }
};

// DFS
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        std::vector<int> preorder;
        preorderDFS(root, preorder);
        return preorder;
    }
private:
    void preorderDFS(TreeNode* node, std::vector<int>& preorder) {
        // note the reference &
        if (node == nullptr) return;
        preorder.push_back(node->val);
        preorderDFS(node->left, preorder);
        preorderDFS(node->right, preorder);
    }
};

File: binary_tree/binary_tree_preorder_traversal.144.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# DFS
# O(n) time:
    # Each node is visited exactly once. 
    # For every non-null node, preorderDFS does a constant amount of work — **one push_back plus two recursive calls** — all O(1). 
    # With n nodes, that's n constant-time operations, giving O(n).
# O(n) space:
    # Space: the depth of the recursion call stack.
    # Worst when degenerates h=n; best when balanced h=log(n).
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        preorder = []
        def preorderDFS(node):
            if node is None:
                return
            preorder.append(node.val)
            preorderDFS(node.left)
            preorderDFS(node.right)
        preorderDFS(root)
        return preorder

# Iterative DFS using an explicit stack
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        preorder = []
        st = [root]
        while st:
            node = st.pop()
            preorder.append(node.val)
            if node.right:
                st.append(node.right)
            if node.left:
                st.append(node.left)
        return preorder